src/Std/Data/DHashMap/Raw.lean

```diff
@@ -0,0 +1,438 @@
+    Array (α × β) :=
+  m.foldl (fun acc k v => acc.push ⟨k, v⟩) #[]
+
+/-- Returns a list of all keys present in the hash map in some order. -/
+@[inline] def keys (m : Raw α β) : List α :=
+  m.foldl (fun acc k _ => k :: acc) []
+
+/-- Returns an array of all keys present in the hash map in some order. -/
+@[inline] def keysArray (m : Raw α β) : Array α :=
+  m.foldl (fun acc k _ => acc.push k) #[]
+
+/-- Returns a list of all values present in the hash map in some order. -/
+@[inline] def values {β : Type v} (m : Raw α (fun _ => β)) : List β :=
+  m.foldl (fun acc _ v => v :: acc) []
+
+/-- Inserts multiple mappings into the hash map by iterating over the given collection and calling
+`insert`. If the same key appears multiple times, the last occurrence takes precendence. -/
+@[inline] def insertMany [BEq α] [Hashable α] {ρ : Type w} [ForIn Id ρ ((a : α) × β a)]
+    (m : Raw α β) (l : ρ) : Raw α β :=
+  if h : 0 < m.buckets.size then
+    (Raw₀.insertMany ⟨m, h⟩ l).1
+  else m -- will never happen for well-formed inputs
+
+@[inline, inherit_doc Raw.insertMany] def Const.insertMany {β : Type v} [BEq α] [Hashable α]
+    {ρ : Type w} [ForIn Id ρ (α × β)] (m : Raw α (fun _ => β)) (l : ρ) : Raw α (fun _ => β) :=
+  if h : 0 < m.buckets.size then
+    (Raw₀.Const.insertMany ⟨m, h⟩ l).1
+  else m -- will never happen for well-formed inputs
+
+@[inline, inherit_doc Raw.insertMany] def Const.insertManyUnit [BEq α] [Hashable α] {ρ : Type w}
+    [ForIn Id ρ α] (m : Raw α (fun _ => Unit)) (l : ρ) : Raw α (fun _ => Unit) :=
+  if h : 0 < m.buckets.size then
+    (Raw₀.Const.insertManyUnit ⟨m, h⟩ l).1
+  else m -- will never happen for well-formed inputs
+
+/-- Creates a hash map from a list of mappings. -/
+@[inline] def ofList [BEq α] [Hashable α] {ρ : Type w} [ForIn Id ρ ((a : α) × β a)] (l : ρ) :
+    Raw α β :=
+  insertMany ∅ l
+
+@[inline, inherit_doc Raw.ofList] def Const.ofList {β : Type v} [BEq α] [Hashable α] {ρ : Type w}
+    [ForIn Id ρ (α × β)] (l : ρ) : Raw α (fun _ => β) :=
+  Const.insertMany ∅ l
+
+@[inline, inherit_doc Raw.ofList] def Const.unitOfList [BEq α] [Hashable α] {ρ : Type w}
+    [ForIn Id ρ α] (l : ρ) : Raw α (fun _ => Unit) :=
+  Const.insertManyUnit ∅ l
+
+/-- Returns the number of buckets in the internal representation of the hash map. This function may
+be useful for things like monitoring system health, but it should be considered an internal
+implementation detail. -/
+def Internal.numBuckets (m : Raw α β) : Nat :=
+  m.buckets.size
+
+instance [Repr α] [(a : α) → Repr (β a)] : Repr (Raw α β) where
+  reprPrec m prec := Repr.addAppParen ("Std.DHashMap.Raw.ofList " ++ reprArg m.toList) prec
+
+end Unverified
+
+section WF
+
+/-- Well-formedness predicate for hash maps. Users of `DHashMap` will not need to interact with
+this. Users of `DHashMap.Raw` will need to provide proofs of `WF` to lemmas and should use lemmas
+like `WF.empty` and `WF.insert` to show that map operations preserve well-formedness. The
```

Comment:
Is the lemma name always the same as the operation name? If so, it'd be good to say so here.

Reply:
Yes, done.